src/Compatibility/Microsoft.DotNet.ApiCompatibility/ErrorSuppression/SuppressionEngine.cs

```diff
@@ -0,0 +1,200 @@
+{
+    /// <summary>
+    /// Collection of Suppressions which is able to add suppressions, check if a specific error is suppressed, and write all suppressions
+    /// down to a file. The engine is thread-safe.
+    /// </summary>
+    public class SuppressionEngine
+    {
+        protected HashSet<Suppression> _validationSuppressions;
+        private readonly ReaderWriterLockSlim _readerWriterLock = new();
+        private readonly XmlSerializer _serializer = new(typeof(Suppression[]), new XmlRootAttribute("Suppressions"));
+
+        protected SuppressionEngine(string suppressionFile)
+        {
+            _validationSuppressions = ParseSuppressionFile(suppressionFile);
+        }
+
+        protected SuppressionEngine()
+        {
+            _validationSuppressions = new HashSet<Suppression>();
+        }
+
+        /// <summary>
+        /// Checks if the passed in error is suppressed or not.
+        /// </summary>
+        /// <param name="diagnosticId">The diagnostic ID of the error to check.</param>
+        /// <param name="target">The target of where the <paramref name="diagnosticId"/> should be applied.</param>
+        /// <param name="left">Optional. The left operand in a APICompat error.</param>
+        /// <param name="right">Optional. The right operand in a APICompat error.</param>
+        /// <returns><see langword="true"/> if the error is already suppressed. <see langword="false"/> otherwise.</returns>
+        public bool IsErrorSuppressed(string? diagnosticId, string? target, string? left = null, string? right = null, bool isBaselineSuppression = false)
+        {
+            var suppressionToCheck = new Suppression()
+            {
+                DiagnosticId = diagnosticId,
+                Target = target,
+                Left = left,
+                Right = right,
+                IsBaselineSuppression = isBaselineSuppression
+            };
+            return IsErrorSuppressed(suppressionToCheck);
+        }
+
+        /// <summary>
+        /// Checks if the passed in error is suppressed or not.
+        /// </summary>
+        /// <param name="error">The <see cref="Suppression"/> error to check.</param>
+        /// <returns><see langword="true"/> if the error is already suppressed. <see langword="false"/> otherwise.</returns>
+        public bool IsErrorSuppressed(Suppression error)
+        {
+            _readerWriterLock.EnterReadLock();
+            try
+            {
+                if (_validationSuppressions.Contains(error))
+                {
+                    return true;
+                }
+                else
+                {
+                    // See if the error is globally suppressed by checking if the same diagnosticid and target are entered
+                    // without any left and right.
+                    return (error.DiagnosticId == null || error.DiagnosticId.StartsWith("cp", StringComparison.InvariantCultureIgnoreCase)) &&
```

Comment:
@joperezr do you remember why you explicitly checked for diagnostics that start with the `cp` prefix? Why would we explicitly block to globally suppress package validation diagnostics?

Reply:
Yes, this is because we only allowed global suppressions for API Compatibility diagnostics that were detected. Package validation has two different types of diagnostics: PKVxxxxx which are the ones caused by an issue with how your package is authored, and CPxxxxxx which are the issues where compatibility problems where detected. We only allowed global suppressions for the latter.

Reply:
What's the underlying reason that you can suppress PKV issues via the NoWarn property or when you have a direct match suppression but not via "global suppressions"?

Reply:
I think that when we did this we wanted to be able to avoid a left and a right, that is what we called "global supression", so basically the support to only have a diagnostic and target and have that apply to all left's and right's. The thing that is special with PKV, is that IIRC they don't have a left and right already. A PKV is just a package violation, but you are not really comparing left and right. Even when you have a pkv diagnostic that is comparing a baseline package to the new one, the diagnostic you would get would still not have left and right, it would only have IsBaselineValidation set to true. So essentially, there is no harm in also doing PKV global suppressions as well, but there is no benefit either since left and right are never a thing for PKV errors, and what global suppression is doing in this next line is to see if there is an error matching without a left and right.

Reply:
That makes perfect sense. Right, package validation doesn't compare a specific left to a specific right. In most cases it actually complains that a specific left or right cannot be found. Thanks for the detailed explanation.

Reply:
I added comments to better describe the existing behavior in my PR: https://github.com/dotnet/sdk/pull/25938.